### PR TITLE
Refactor HTTP module and package structure

### LINKED ISSUE #17279: Improve package structure to put scala and java types on same level
In the `akka-http-core` and `akka-http` modules the chosen solution for limiting code duplication is the "scala-side-extends-java-side" approach.

Thereby the current package structure doesn't put scala and java on the same level. For example, right now we have:

Scala | Java
----- | ----
akka.http.model.HttpEntity | akka.http.model.japi.HttpEntity
akka.http.model.headers.HttpChallenge | akka.http.model.japi.headers.HttpChallenge
akka.http.Http | akka.http.japi.Http

The proposal is to introduce two super packages `akka.http.scala` and `akka.http.java` under which we create identical sub-packages for all public types that have representations in both languages. So the above example would become:

Scala | Java
----- | ----
akka.http.scala.model.HttpEntity | akka.http.java.model.HttpEntity
akka.http.scala.model.headers.HttpChallenge | akka.http.java.model.headers.HttpChallenge
akka.http.scala.Http | akka.http.java.Http

Internal (implementation) types would only exist in their own side (mostly scala).

/cc @jrudolph

### PR DESCRIPTION
**Pull request by sirthias**
Closes #17279.

/cc @jrudolph 

(This PR is in progress. I'll keep pushing commits in the next minutes to get it completed.
You can already base stuff in `htc` on this though, if required.)

### CODE REVIEW

File: akka-http-core/src/main/scala/akka/http/impl/util/package.scala [hunk 1]
```diff
@@ -120,16 +120,7 @@ package object util {
 
   private[this] val _identityFunc: Any ⇒ Any = x ⇒ x
   /** Returns a constant identity function to avoid allocating the closure */
-  def identityFunc[T]: T ⇒ T = _identityFunc.asInstanceOf[T ⇒ T]
-
-  def humanReadableByteCount(bytes: Long, si: Boolean): String = {
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/impl/util/package.scala]: was not used?

**Comment by drewhk** [comment]
Cannot show the entire diff, but LGTM so far

File: akka-http-core/src/main/scala/akka/http/impl/util/package.scala (hunk 1 above)
Comment (sirthias) [inline akka-http-core/src/main/scala/akka/http/impl/util/package.scala]: Nope. Apparently not.

**Comment by jrudolph** [comment]
As this doesn't merge even now, how are we going to proceed? Do you want to rebase, @sirthias, or do we add a merge commit in the end?

**Comment by sirthias** [comment]
I'm on it. The teams keeps merging stuff in release-2.3-dev and somehow this conflict withs everything.
I'll just rebased a 3rd time, let's see.

**Comment by sirthias** [comment]
Damn. Endre already merged your PR.

**Comment by drewhk** [comment]
Aah, sorry

**Comment by patriknw** [comment]
we'll hold back any more merging now, until you are done, sorry

**Comment by jrudolph** [comment]
Maybe we just merge it in the end this time?

**Comment by patriknw** [comment]
that is ok!

**Comment by patriknw** [comment]
I mean, we are done with all other PRs now (I think)

**Comment by jrudolph** [comment]
@patriknw and I meant we accept the merge conflict and solve it by either merging manually in the end or adding a separate merge commit as the last one to deal with stuff directly before merging the PR.

**Comment by sirthias** [comment]
Ok, wasn't so bad after all.
So this thing *now* contains the package refactoring for `akka-http-core`, which was the hardest I think.
I'll add commit for the other akka-http modules now.

**Comment by jrudolph** [comment]
Nice!

**Comment by patriknw** [comment]
Separate merge commit as the last one is ok - and that was what I was saying ok to first, but it could probably have been interpreted in different ways. Thanks.

**Comment by jrudolph** [comment]
@patriknw ok :)

**Comment by sirthias** [comment]
Ok, finally everything is done.
Sorry it took so long.
Was a real beast.

**Comment by patriknw** [comment]
great, then we wait for validation

**Comment by sirthias** [comment]
Somehow this shuffling removed 900 LOC, strange.
I wasn't actively deleting anything major...

**Comment by sirthias** [comment]
Ah, sorry, I was misreading githubs stats info on insertions/deletions as LOC.
Nevermind.

**Comment by sirthias** [comment]
Latest validation failure looks like a timing issue.
Just rebased once more and pushed again.
Let's see whether we've got more luck this time.

**Comment by patriknw** [comment]
we can also merge this, and validate it one more time in @jrudolph PR?

**Comment by sirthias** [comment]
Yes, that would work just as well I guess.

**Comment by 2m** [comment]
Okay then. Merging this and cancelling validation job.

**Pull request merged**